### PR TITLE
ipset: Rework the reconciler to use batch ops

### LINKED ISSUE #31537: ipset entries transitory drop on agent restart due to the new reconcile logic
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

https://github.com/cilium/cilium/pull/31099 introduced a new reconciler logic to manage the ipsets related to the Node IPv4 and IPv6 addresses, which are used to avoid maquerading pod to remote node traffic when Cilium is configured in native routing mode. Differently from the previous implementation, it ensures that the ipset entries match the expected ones, possibly removing stale ones while doing upsert operations.

However, this approach causes the temporary removal of existing and valid entries on restart, when the full list of remote nodes has not been yet retrieved. In turn, this would possibly disrupt pod to remote traffic, which would be incorrectly masqueraded during this time frame. In particular, this applies when host firewall is in use, as the remote node would no longer be able to associate the correct identity with the traffic flow.

A possible solution could be avoiding to perform any removal of possible stale entries until a full synchronization (from either k8s or the kvstore for the local cluster, and all remote clusters if clustermesh is enabled) has been completed. At that point, a full synchronization should be triggered to enforce the cleanup of remaining stale entries (which at this point are guaranteed to be actually stale), and finally normal operations can be resumed.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by pippolo84**
Rework the ipset table to store a single address per entry. Consequently, rewrite the ipset reconciler to:

- use batch operations and `ipset restore`
- delete addresses only in DeleteBatch and Prune operations

Finally, introduce the concept of an ipset manager initializer, to delay the addresses pruning during agent restart, before the initial nodes listing from k8s or kvstore is completed.

Refer to the individual commit messages and the related issue for further details.

Fixes: https://github.com/cilium/cilium/issues/31537

### CODE REVIEW

File: pkg/datapath/iptables/ipset/ipset.go [hunk 1]
```diff
@@ -98,33 +81,41 @@ func (m *manager) RemoveFromIPSet(name string, addrs ...netip.Addr) {
...
-	lc cell.Lifecycle,
-	logger logrus.FieldLogger,
-	db *statedb.DB,
-	table statedb.RWTable[*tables.IPSet],
-	cfg config,
+	p params,
 	ipset *ipset,
 	_ reconciler.Reconciler[*tables.IPSet], // needed to enforce the correct hive ordering
-) Manager {
-	db.RegisterTable(table)
```
Comment (joamaki) [inline pkg/datapath/iptables/ipset/ipset.go:110]:
```
func (m *manager) NewHandle() Handle {
  if m.started.Load() {
    panic(...)
  }
  m.startWG.Add(1)
  ...
}
// onStart is a oneshot async func
func (m *manager) onStart() {
  m.startWG.Wait() // wait for all handles to do initial population
  m.reconcilerLifecycle.Start(ctx) // start the reconciler
}

func (h *handle) InitDone() {
  h.manager.startWG.Done()
}
type Handle interface {
  AddToIPSet(...)
  InitDone()
}
```

File: pkg/datapath/iptables/ipset/cell.go [hunk 2]
```diff
@@ -60,7 +63,7 @@ func newReconcilerConfig(
 	ops reconciler.Operations[*tables.IPSetEntry],
 ) reconciler.Config[*tables.IPSetEntry] {
 	return reconciler.Config[*tables.IPSetEntry]{
-		FullReconcilationInterval: 10 * time.Second,
+		FullReconcilationInterval: 10 * time.Minute,
```
Comment (pippolo84) [inline pkg/datapath/iptables/ipset/cell.go]: Increase to 30 min

File: pkg/datapath/tables/ipset.go [hunk 3]
```diff
@@ -5,91 +5,65 @@ package tables
...
-	"strings"
 
-	"github.com/cilium/cilium/pkg/container"
 	"github.com/cilium/cilium/pkg/statedb"
 	"github.com/cilium/cilium/pkg/statedb/index"
 	"github.com/cilium/cilium/pkg/statedb/reconciler"
 )
 
-const IPSetTableName = "ipset"
+const IPSetEntryTableName = "ipset-entry"
```
Comment (pippolo84) [inline pkg/datapath/tables/ipset.go]: Change to "ipsets"

**Comment by pippolo84** [comment]
A performance test done by @giorio94 against the previous implementation (based on a multiple addresses per table entry model) highlighted that the `ipset list` operation can become quite expensive in large multi-cluster setup in terms of memory allocation, leading to a high GC pressure. Here the related pprof:

pprof

The current implementation already addresses this partially, since moving the table to a single address per entry model simplifies the `Update` implementation, removing the need for relying on `ipset list` in favor of `ipset create`.
To be on the safe side, the [last commit](https://github.com/cilium/cilium/pull/31638/commits/c19d61d5c9787070a49e955e4ba9f6c980fa28d9) updates the implementation of the `Delete` operation too, so that the entire incremental reconciliation path will not use the `ipset list` anymore.

File: pkg/datapath/iptables/ipset/ops.go [hunk 4]
```diff
@@ -103,19 +135,16 @@ func reconcile(
...
+		return fmt.Errorf("unable to delete from ipset: %w", err)
 	}
 
 	toAdd := desired.Difference(curSet)
-	for addr := range toAdd {
-		if err := ipset.add(ctx, name, addr); err != nil {
-			return fmt.Errorf("unable to add addr %s from ipset %s: %w", name, addr, err)
-		}
+	addBatch := map[string][]netip.Addr{name: toAdd.UnsortedList()}
+	if err := ipset.delBatch(ctx, addBatch); err != nil {
```
Comment (pippolo84) [inline pkg/datapath/iptables/ipset/ops.go]: Should be `ipset.addBatch`

File: pkg/datapath/iptables/ipset/ipset.go [hunk 5]
```diff
@@ -263,16 +262,43 @@ func (i *ipset) add(ctx context.Context, name string, addr netip.Addr) error {
...
+	return err
+}
+
 func (i *ipset) run(ctx context.Context, args ...string) ([]byte, error) {
 	i.log.Debugf("Running command %s", i.fullCommand(args...))
-	return i.exec(ctx, "ipset", args...)
+	return i.exec(ctx, "ipset", "", args...)
+}
+
+func (i *ipset) runStdin(ctx context.Context, stdin string, args ...string) ([]byte, error) {
```
Comment (pippolo84) [inline pkg/datapath/iptables/ipset/ipset.go]: This seems unused

**Comment by pippolo84** [comment]
@joamaki contributed a change to the reconciler logic to coalesce multiple additions and further improve the performance in case of node churn. Specifically, we now relies on batch operations and on the `ipset restore` subcommand.
The performance are now comparable to what we had before the introduction of the reconciler for ip sets.

Updated profiles for CPU and memory (alloc_space samples) in a cluster mesh setup:

cpu

mem

I've updated the title of the PR accordingly.

**Review by joamaki — APPROVED** [review]

File: pkg/datapath/iptables/ipset/ipset.go [hunk 6]
```diff
@@ -156,6 +188,11 @@ func newIPSetManager(
 }
 
 func (m *manager) init(ctx context.Context, _ cell.HealthReporter) error {
+	// wait for all initializers to complete before finalizing manager
+	// initialization and allowing prune operations in the ipset reconciler
+	m.started.Store(true)
+	m.startWG.Wait()
```
Comment (giorio94) [inline pkg/datapath/iptables/ipset/ipset.go]: I guess this could be moved below the ipset cleanup step, as that's independent of node synchronization.

File: pkg/node/manager/manager.go [hunk 7]
```diff
@@ -827,6 +829,13 @@ func (m *manager) NodeDeleted(n nodeTypes.Node) {
 	}
 }
 
+// NodeSync signals the manager that the initial nodes listing (either from k8s or kvstore)
+// has been completed. This allows the manager to initiate the deletion of possible stale entries
+// in the kernel IP sets managed by Cilium.
+func (m *manager) NodeSync() {
+	m.ipsetInitializer.InitDone()
+}
```
Comment (giorio94) [inline pkg/node/manager/manager.go:837]: `NodeSync` may be called twice if the kvstore is enabled (once when the informer syncs, the other when the kvstore syncs). However, `InitDone` is implemented on top of a wait group, which panics if it turns negative.

Comment (giorio94) [reply pkg/node/manager/manager.go:837]: Additionally, this logic is not accounting for the clustermesh case. You can retrieve whether clustermesh is synchronized using `(*ClusterMesh).NodesSynced()`.

File: pkg/datapath/iptables/ipset/ops.go [hunk 8]
```diff
@@ -30,37 +31,61 @@ type ops struct {
 	ipset   *ipset
 }
 
-func (ops *ops) Update(ctx context.Context, _ statedb.ReadTxn, entry *tables.IPSetEntry, changed *bool) error {
+// DeleteBatch implements reconciler.BatchOperations.
+func (ops *ops) DeleteBatch(ctx context.Context, txn statedb.ReadTxn, batch []reconciler.BatchEntry[*tables.IPSetEntry]) {
```
Comment (giorio94) [inline pkg/datapath/iptables/ipset/ops.go]: Nit. It seems a bit more natural if UpdateBatch comes before DeleteBatch (and the diff would also improve a bit).

File: pkg/datapath/iptables/ipset/ops.go [hunk 9]
```diff
@@ -30,37 +31,61 @@ type ops struct {
...
+		for i := range batch {
+			batch[i].Result = err
+		}
 	}
+}
 
-	if err := ops.ipset.add(ctx, entry.Name, entry.Addr); err != nil {
-		return err
+// UpdateBatch implements reconciler.BatchOperations.
+func (ops *ops) UpdateBatch(ctx context.Context, txn statedb.ReadTxn, batch []reconciler.BatchEntry[*tables.IPSetEntry]) {
```
Comment (giorio94) [inline pkg/datapath/iptables/ipset/ops.go]: Differently from before, both `UpdateBatch` and `DeleteBatch` no longer check whether the ipset actually exists before attempting the operations. It looks reasonable to me, especially performance-wise, just making sure that the change was intentional.

File: pkg/statedb/reconciler/reconciler.go [hunk 10]
```diff
@@ -107,12 +107,16 @@ func WaitForReconciliation[Obj any](ctx context.Context, db *statedb.DB, table s
 }
 
 func (r *reconciler[Obj]) loop(ctx context.Context, health cell.HealthReporter) error {
+
```
Comment (giorio94) [inline pkg/statedb/reconciler/reconciler.go]: Nit. Should be likely reverted.

File: pkg/statedb/reconciler/reconciler.go [hunk 11]
```diff
@@ -107,12 +107,16 @@ func WaitForReconciliation[Obj any](ctx context.Context, db *statedb.DB, table s
...
 	var fullReconTickerChan <-chan time.Time
 	if r.Config.FullReconcilationInterval > 0 {
 		fullReconTicker := time.NewTicker(r.Config.FullReconcilationInterval)
 		defer fullReconTicker.Stop()
 		fullReconTickerChan = fullReconTicker.C
 	}
+	if r.Config.RateLimiter != nil {
+		defer r.Config.RateLimiter.Stop()
+	}
```
Comment (giorio94) [inline pkg/statedb/reconciler/reconciler.go:118]: Nit. May be worth to move it to a separate commit, as it appears an unrelated fix.

**Review by giorio94 — CHANGES_REQUESTED** [review]
Thanks!

A couple of comments inline about waiting for clustermesh synchronization as well, and a possible panic due to double sync. And a few minor nits.

Concerning the node synchronization logic, it may be worth considering that we'll likely need something very similar for WireGuard to address https://github.com/cilium/cilium/issues/31985, to ensure that the current logic is extensible enough to address that case as well.

File: pkg/datapath/iptables/ipset/ops.go (hunk 9 above)
Comment (pippolo84) [reply pkg/datapath/iptables/ipset/ops.go]: Yep, that is intentional and it is meant to further improve the performance removing the exec to run `ipset create`. In case of removed sets, we should be able to recover with ta full reconciliation.

File: pkg/node/manager/manager.go (hunk 7 above)
Comment (pippolo84) [reply pkg/node/manager/manager.go:837]: Nice catch! Switched to an int counter with a `sync.Cond`, so we can check for counter to be equal to 0 or negative. Also, added an Invoke in clustermesh cell to wait for external nodes sync.

Comment (giorio94) [reply pkg/node/manager/manager.go:837]: Hmm, I only had a quick look, but I'm still not convinced this would work as expected.

In the the kvstore + clustermesh case, the following sequence may occur:

* Node manager + clustermesh take the initializer -> counter = 2
* CRD synchronization -> InitDone -> counter = 1
* kvstore synchronization -> InitDone -> counter = 0
* The prune operation is executed, even though clustermesh has not yet synchronized. 

I think we should either make sure that `InitDone` is called exactly once, or make it idempotent (i.e., a no-op after the first execution).

As a side note, you could use a StoppableWaitGroup instead of the custom counter + `sync.Cond` logic.

Comment (pippolo84) [reply pkg/node/manager/manager.go:837]: Yeah, the fact that we can flip back and forth from kvstore to crd is not playing nice with the initializers logic.

> I think we should either make sure that InitDone is called exactly once, or make it idempotent (i.e., a no-op after the first execution).

I think the first requires an ad-hoc solution for the kvstore/crd case, it is probably better to make the `InitDone` idempotent, so to keep having one "signal" for each "initializer" taken from the manager.

I'll try with the `StoppableWaitGroup`, thanks for the heads up!

File: daemon/cmd/status_test.go [hunk 12]
```diff
@@ -36,14 +38,36 @@ var (
...
+
 func (g *GetNodesSuite) SetUpTest(c *C) {
 	option.Config.IPv4ServiceRange = AutoCIDR
 	option.Config.IPv6ServiceRange = AutoCIDR
 }
 
 func (g *GetNodesSuite) SetUpSuite(c *C) {
 	var err error
-	nm, err = manager.New(fakeConfig, nil, nil, nil, manager.NewNodeMetrics(), cell.TestScope())
+	nm, err = manager.New(fakeConfig, nil, newFakeIPSetMock(), nil, manager.NewNodeMetrics(), cell.TestScope())
```
Comment (giorio94) [inline daemon/cmd/status_test.go]: Any reason for not using the fake type in `pkg/datapath/fake/types/ipset.go`?

File: pkg/datapath/iptables/ipset/ipset.go [hunk 13]
```diff
@@ -50,17 +51,13 @@ type Initializer interface {
 }
 
 type initializer struct {
-	cond    *sync.Cond
-	counter *int
+	once sync.Once
+	wg   *lock.StoppableWaitGroup
```
Comment (giorio94) [inline pkg/datapath/iptables/ipset/ipset.go:55]: Should be moved to the `iptables/ipset: Add initializer to delay entries deletion` commit. The same applies for the other changes in this file.

File: pkg/clustermesh/ipsetsyncer.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+	jg := jobRegistry.NewGroup(
+		scope,
+		job.WithLogger(logger),
+		job.WithPprofLabels(pprof.Labels("cell", "clustermesh-ipset-syncer")),
+	)
+	jg.Add(job.OneShot("clustermesh-ipset-syncer", func(ctx context.Context, _ cell.HealthReporter) error {
+		// wait for initial nodes listing from all remote clusters
+		// before allowing stale ipset entries deletion
+		cm.NodesSynced(ctx)
```
Comment (giorio94) [inline pkg/clustermesh/ipsetsyncer.go]: Nit. It is likely better to check the possible error, and return in that case. The only possible error is if the context gets closed, but in that case we don't want to accidentally trigger the pruning step due to race conditions (although theoretically the corresponding context should be canceled as well).

**Review by giorio94 — COMMENTED** [review]
Thanks! A couple of minor comments, then I'll do the final pass.

File: pkg/datapath/iptables/ipset/ipset.go (hunk 13 above)
Comment (pippolo84) [reply pkg/datapath/iptables/ipset/ipset.go:55]: Oops, I've messed up the rebase. I'm gonna fix it.

Comment (giorio94) [reply pkg/datapath/iptables/ipset/ipset.go:55]: Thanks!

File: pkg/clustermesh/ipsetsyncer.go (hunk 14 above)
Comment (pippolo84) [reply pkg/clustermesh/ipsetsyncer.go:39]: I'll specifically check for the `context.Canceled` in order to avoid re-triggering the job :+1:

File: daemon/cmd/status_test.go (hunk 12 above)
Comment (pippolo84) [reply daemon/cmd/status_test.go]: I thought we will end up in an import loop but it seems I was wrong. Fixed.

**Review by giorio94 — APPROVED** [review]
Thanks!
